File: src/commands/hook-install.ts
import { Command } from 'commander';
import { existsSync, mkdirSync, readFileSync, writeFileSync, chmodSync } from 'fs';
import { join } from 'path';
import { execSync } from 'child_process';

export function hookInstallCommand(): Command {
  return new Command('hook-install')
    .description('Install Claude Code hooks for mm integration')
    .option('--dry-run', 'show what would be written without writing')
    .option('--precommit', 'also install git pre-commit hook for claim conflict detection')
    .action((options, cmd) => {
      const projectDir = process.env.CLAUDE_PROJECT_DIR || process.cwd();
      const claudeDir = join(projectDir, '.claude');
      const settingsPath = join(claudeDir, 'settings.local.json');

      const hooksConfig = {
        hooks: {
          SessionStart: [
            {
              matcher: 'startup',
              hooks: [
                {
                  type: 'command',
                  command: 'mm hook-session startup',
                  timeout: 10,
                },
              ],
            },
            {
              matcher: 'resume',
              hooks: [
                {
                  type: 'command',
                  command: 'mm hook-session resume',
                  timeout: 10,
                },
              ],
            },
          ],
          UserPromptSubmit: [
            {
              hooks: [
                {
                  type: 'command',
                  command: 'mm hook-prompt',
                  timeout: 5,
                },
              ],
            },
          ],
        },
      };

      if (options.dryRun) {
        console.log('Would write to:', settingsPath);
        console.log(JSON.stringify(hooksConfig, null, 2));
        return;
      }

      // Ensure .claude directory exists
      if (!existsSync(claudeDir)) {
        mkdirSync(claudeDir, { recursive: true });
      }

      // Merge with existing settings if present
      let existingSettings: Record<string, unknown> = {};
      if (existsSync(settingsPath)) {
        try {
          existingSettings = JSON.parse(readFileSync(settingsPath, 'utf-8'));
        } catch {
          // Ignore parse errors, start fresh
        }
      }

      // Merge hooks config
      const merged = {
        ...existingSettings,
        hooks: {
          ...(existingSettings.hooks as Record<string, unknown> || {}),
          ...hooksConfig.hooks,
        },
      };

      writeFileSync(settingsPath, JSON.stringify(merged, null, 2) + '\n');
      console.log(`Hooks installed to ${settingsPath}`);
      console.log('');
      console.log('Installed hooks:');
      console.log('  SessionStart (startup/resume) - prompts agent registration or injects context');
      console.log('  UserPromptSubmit - injects room messages and @mentions before each prompt');

      // Install git pre-commit hook if requested
      if (options.precommit) {
        installPrecommitHook(projectDir, options.dryRun);
      }

      console.log('');
      console.log('Restart Claude Code to activate hooks.');
    });
}

function installPrecommitHook(projectDir: string, dryRun: boolean): void {
  // Find git root
  let gitRoot: string;
  try {
    gitRoot = execSync('git rev-parse --show-toplevel', {
      encoding: 'utf-8',
      cwd: projectDir,
      stdio: ['pipe', 'pipe', 'pipe'],
    }).trim();
  } catch {
    console.log('');
    console.log('⚠️  Not in a git repository, skipping pre-commit hook installation');
    return;
  }

  const hooksDir = join(gitRoot, '.git', 'hooks');
  const precommitPath = join(hooksDir, 'pre-commit');

  const hookScript = `#!/bin/sh
# mm pre-commit hook - detects file claim conflicts
# Installed by: mm hook-install --precommit

mm hook-precommit
`;

  if (dryRun) {
    console.log('');
    console.log('Would write git pre-commit hook to:', precommitPath);
    console.log(hookScript);
    return;
  }

  // Check if pre-commit hook already exists
  if (existsSync(precommitPath)) {
    const existing = readFileSync(precommitPath, 'utf-8');
    if (existing.includes('mm hook-precommit')) {
      console.log('');
      console.log('Git pre-commit hook already installed');
      return;
    }

    // Append to existing hook
    const updated = existing.trimEnd() + '\n\n# mm file claim conflict detection\nmm hook-precommit\n';
    writeFileSync(precommitPath, updated);
    console.log('');
    console.log(`Added mm hook to existing pre-commit hook at ${precommitPath}`);
    return;
  }

  // Ensure hooks directory exists
  if (!existsSync(hooksDir)) {
    mkdirSync(hooksDir, { recursive: true });
  }

  // Write new hook
  writeFileSync(precommitPath, hookScript);
  chmodSync(precommitPath, '755');
  console.log('');
  console.log(`Git pre-commit hook installed at ${precommitPath}`);
  console.log('  Warns on file claim conflicts when committing');
}

File: internal/command/hooks/hook_install.go
package hooks

import (
	"encoding/json"
	"fmt"
	"io"
	"os"
	"os/exec"
	"path/filepath"
	"strings"

	"github.com/spf13/cobra"
)

type hookSettings struct {
	Hooks map[string]any `json:"hooks,omitempty"`
}

// NewHookInstallCmd installs Claude Code hooks for fray integration.
func NewHookInstallCmd() *cobra.Command {
	cmd := &cobra.Command{
		Use:   "hook-install",
		Short: "Install Claude Code hooks for fray integration",
		RunE: func(cmd *cobra.Command, args []string) error {
			dryRun, _ := cmd.Flags().GetBool("dry-run")
			precommit, _ := cmd.Flags().GetBool("precommit")

			projectDir := os.Getenv("CLAUDE_PROJECT_DIR")
			if projectDir == "" {
				cwd, err := os.Getwd()
				if err != nil {
					return writeCommandError(cmd, err)
				}
				projectDir = cwd
			}

			claudeDir := filepath.Join(projectDir, ".claude")
			settingsPath := filepath.Join(claudeDir, "settings.local.json")

			hooksConfig := hookSettings{
				Hooks: map[string]any{
					"SessionStart": []any{
						map[string]any{
							"matcher": "startup",
							"hooks": []any{map[string]any{
								"type":    "command",
								"command": "fray hook-session startup",
								"timeout": 10,
							}},
						},
						map[string]any{
							"matcher": "resume",
							"hooks": []any{map[string]any{
								"type":    "command",
								"command": "fray hook-session resume",
								"timeout": 10,
							}},
						},
					},
					"UserPromptSubmit": []any{
						map[string]any{
							"hooks": []any{map[string]any{
								"type":    "command",
								"command": "fray hook-prompt",
								"timeout": 5,
							}},
						},
					},
					"PreCompact": []any{
						map[string]any{
							"hooks": []any{map[string]any{
								"type":    "command",
								"command": "fray hook-precompact",
								"timeout": 5,
							}},
						},
					},
				},
			}

			out := cmd.OutOrStdout()

			if dryRun {
				fmt.Fprintf(out, "Would write to: %s\n", settingsPath)
				data, _ := json.MarshalIndent(hooksConfig, "", "  ")
				fmt.Fprintln(out, string(data))
				if precommit {
					installPrecommitHook(projectDir, true, out)
				}
				return nil
			}

			if err := os.MkdirAll(claudeDir, 0o755); err != nil {
				return writeCommandError(cmd, err)
			}

			settings := hookSettings{}
			if data, err := os.ReadFile(settingsPath); err == nil {
				_ = json.Unmarshal(data, &settings)
			}
			if settings.Hooks == nil {
				settings.Hooks = map[string]any{}
			}

			for key, value := range hooksConfig.Hooks {
				settings.Hooks[key] = value
			}

			merged, err := json.MarshalIndent(settings, "", "  ")
			if err != nil {
				return writeCommandError(cmd, err)
			}
			merged = append(merged, '\n')
			if err := os.WriteFile(settingsPath, merged, 0o644); err != nil {
				return writeCommandError(cmd, err)
			}

			fmt.Fprintf(out, "Hooks installed to %s\n\n", settingsPath)
			fmt.Fprintln(out, "Installed hooks:")
			fmt.Fprintln(out, "  SessionStart (startup/resume) - prompts agent registration or injects context")
			fmt.Fprintln(out, "  UserPromptSubmit - injects room messages and @mentions before each prompt")
			fmt.Fprintln(out, "  PreCompact - reminds to preserve work before context compaction")

			if precommit {
				installPrecommitHook(projectDir, false, out)
			}

			fmt.Fprintln(out, "")
			fmt.Fprintln(out, "Restart Claude Code to activate hooks.")
			return nil
		},
	}

	cmd.Flags().Bool("dry-run", false, "show what would be written without writing")
	cmd.Flags().Bool("precommit", false, "also install git pre-commit hook for claim conflict detection")

	return cmd
}

func installPrecommitHook(projectDir string, dryRun bool, outWriter io.Writer) {
	gitRoot, err := gitRootDir(projectDir)
	if err != nil {
		fmt.Fprintln(outWriter, "")
		fmt.Fprintln(outWriter, "WARNING: Not in a git repository, skipping pre-commit hook installation")
		return
	}

	hooksDir := filepath.Join(gitRoot, ".git", "hooks")
	precommitPath := filepath.Join(hooksDir, "pre-commit")

	hookScript := strings.Join([]string{
		"#!/bin/sh",
		"# fray pre-commit hook - detects file claim conflicts",
		"# Installed by: fray hook-install --precommit",
		"",
		"fray hook-precommit",
		"",
	}, "\n")

	if dryRun {
		fmt.Fprintln(outWriter, "")
		fmt.Fprintf(outWriter, "Would write git pre-commit hook to: %s\n", precommitPath)
		fmt.Fprintln(outWriter, hookScript)
		return
	}

	if data, err := os.ReadFile(precommitPath); err == nil {
		if strings.Contains(string(data), "fray hook-precommit") {
			fmt.Fprintln(outWriter, "")
			fmt.Fprintln(outWriter, "Git pre-commit hook already installed")
			return
		}
		updated := strings.TrimRight(string(data), "\n") + "\n\n# fray file claim conflict detection\nmm hook-precommit\n"
		if err := os.WriteFile(precommitPath, []byte(updated), 0o755); err != nil {
			fmt.Fprintln(outWriter, "")
			fmt.Fprintf(outWriter, "Failed to update pre-commit hook: %v\n", err)
			return
		}
		fmt.Fprintln(outWriter, "")
		fmt.Fprintf(outWriter, "Added fray hook to existing pre-commit hook at %s\n", precommitPath)
		return
	}

	if err := os.MkdirAll(hooksDir, 0o755); err != nil {
		fmt.Fprintln(outWriter, "")
		fmt.Fprintf(outWriter, "Failed to create hooks directory: %v\n", err)
		return
	}

	if err := os.WriteFile(precommitPath, []byte(hookScript), 0o755); err != nil {
		fmt.Fprintln(outWriter, "")
		fmt.Fprintf(outWriter, "Failed to write pre-commit hook: %v\n", err)
		return
	}
	_ = os.Chmod(precommitPath, 0o755)

	fmt.Fprintln(outWriter, "")
	fmt.Fprintf(outWriter, "Git pre-commit hook installed at %s\n", precommitPath)
	fmt.Fprintln(outWriter, "  Warns on file claim conflicts when committing")
}

func gitRootDir(startDir string) (string, error) {
	cmd := exec.Command("git", "rev-parse", "--show-toplevel")
	cmd.Dir = startDir
	output, err := cmd.Output()
	if err != nil {
		return "", err
	}
	return strings.TrimSpace(string(output)), nil
}
